### PR TITLE
proposal: Deprecate/Remove gossip clustering for Thanos.

### PR DESCRIPTION
**Pull request by Bplotka**
This is quite a major change, so please give us feedback!

Rendered view: https://github.com/improbable-eng/thanos/blob/gossip-remove-proposal/docs/proposals/gossip-removal.md

PTAL: @brancz @povilasv @fabxc @domgreen and others.

@mxinden @stuartnelson3 this may interest you.

### CODE REVIEW

File: docs/proposals/gossip-removal.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+* Gossip operates per `peer` level and there is no way you can abstract multiple peers behind loadbalancer. This hides easy
+solutions from our eyes, e.g how to make Store Gateway HA. Without gossip, you can just use Kubernetes HA Service or any other loadbalancer. 
+To support Store Gateway HA for gossip we would end up implementing LB logic in Thanos Querier (like proposed [here](https://github.com/improbable-eng/thanos/pull/404))
+* At some point we want to be flexible and allow other discovery mechanisms. Gossip does not work for everyone and static flags
+are too.. static. (: We need [File SD](https://github.com/improbable-eng/thanos/issues/492) for flexibility anyway.
+* One thing less to maintain.
+
+## Goals
+
+* Deprecate gossip protocol or remove it immediately (TODO: Gather input here!)
```
[the remark this answers is not in the document]
Comment (brancz) [reply docs/proposals/gossip-removal.md]: My personal take is that Thanos is a very young project and this change is overall for the better so I think a hard break here is fine.

[the remark this answers is not in the document]
Comment (domgreen) [reply docs/proposals/gossip-removal.md]: I tend to agree, publicly declare deprication and the plan as we work on file-sd and then remove once file-sd has been proven out fully.
Users can always take old images from docker hub if needed whilst we move forward.

File: docs/proposals/gossip-removal.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+## Proposal: Steps
+
+* Add File Service Discovery (SD): https://github.com/improbable-eng/thanos/issues/492
+* Remove gossip from the documentation, be clear what talks with what (!)
+* Add "Deprecate" to all places of implementation and flag
+  * Alternatively just remove it? (we are still in RC)
+
+### Backwards compatibility
+
```
[the remark this answers is not in the document]
Comment (Bplotka) [reply docs/proposals/gossip-removal.md]: Interesting idea!

**Comment by povilasv** [comment]
:+1: Great proposal. I think if we can avoid gossip, we should definitely do that.

File: docs/proposals/gossip-removal.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bplotka) [reply docs/proposals/gossip-removal.md]: Updated doc accordingly

File: docs/proposals/gossip-removal.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  * We are still RC, so technically there are no API guarantees yet.
+* Leave --store flags
+* Make sure [File SD](https://github.com/improbable-eng/thanos/issues/492) is in place and documented before removal.
+
+## Proposal: Steps
+
+* Add File Service Discovery (SD): https://github.com/improbable-eng/thanos/issues/492
+* Remove gossip from the documentation, be clear what talks with what (!)
+* Deprecate gossip in code.
+* Remove gossip code and flags AFTER some time (month?)
```
Comment (brancz) [inline docs/proposals/gossip-removal.md]: I don't think there is any time to wait. Once file discovery is implemented and documented it should just be removed directly.

Comment (Bplotka) [reply docs/proposals/gossip-removal.md]: Cool.

@ivan-valkov you were asking how blocking File SD issue is. (:

Comment (domgreen) [reply docs/proposals/gossip-removal.md]: I think we are all in agreement that we should move forward with this approach :+1:

**Comment by brancz** [comment]
:+1:

**Comment by Bplotka** [comment]
Moved status to `approved` and added requirement on FileSD before gossip implementation removal.

**Pull request merged**